src/crs.rs

```diff
@@ -0,0 +1,429 @@
+        // and so we iterate through the substrings backwards collecting
+        // digits and adding them to our EPSG code
+        let mut epsg_code = 0;
+        let mut code_has_started = false;
+        let mut power = 1;
+        // the 10 last bytes should be enough (with a small margin)
+        // as the code is 4 or 5 digits starting at the 2 or 3 byte from the back
+        for byte in piece.iter().rev().take(10) {
+            // if the byte is an ASCII encoded digit
+            if byte.is_ascii_digit() {
+                // mark that the EPSG code has started
+                // so that we can break when we no
+                // longer find digits
+                code_has_started = true;
+
+                // translate from ASCII to digits
+                // and multiply by powers of 10
+                // sum it to build the EPSG
+                // code digit by digit
+                epsg_code += power * (byte - 48) as u16;
+                power *= 10;
+            } else if code_has_started {
+                // we no longer see digits
+                // so the code must be over
+                break;
+            }
+        }
+        if EPSG_RANGE.contains(&epsg_code) {
+            epsg[pi] = Some(epsg_code);
+        }
+    }
+    if let Some(horizontal) = epsg[0] {
+        Ok(EpsgCrs {
+            horizontal,
+            vertical: epsg[1],
+        })
+    } else {
+        Err(Error::UnreadableWktCrs)
+    }
+}
+
+/// Get the EPSG code(s) from GeoTiff-CRS-data
+pub fn get_epsg_from_geotiff_crs(geotiff_crs_data: &GeoTiffCrs) -> Result<EpsgCrs> {
+    let mut out = (0, None);
+    for entry in geotiff_crs_data.entries.iter() {
+        match entry.id {
+            // 2048 and 3072 should not co-exist, but might both be combined with 4096
+            // 1024 should always exist
+            1024 => match &entry.data {
+                GeoTiffData::U16(0) => return Err(Error::UnreadableGeoTiffCrs),
+                GeoTiffData::U16(1) => (), // projected crs
+                GeoTiffData::U16(2) => (), // geographic crs
+                GeoTiffData::U16(3) => (), // geographic + a vertical crs
+                GeoTiffData::U16(32_767) => return Err(Error::UserDefinedCrs),
+                _ => return Err(Error::UnimplementedForGeoTiffStringAndDoubleData),
+            },
+            2048 | 3072 => {
+                if let GeoTiffData::U16(v) = entry.data {
+                    out.0 = v;
+                }
+            }
+            4096 => {
+                // vertical crs
+                if let GeoTiffData::U16(v) = entry.data {
+                    out.1 = Some(v);
+                }
+            }
+            _ => (), // the rest are descriptions and units.
+        }
+    }
+    if !EPSG_RANGE.contains(&out.0) {
+        return Err(Error::UnreadableGeoTiffCrs);
+    }
+    Ok(EpsgCrs {
+        horizontal: out.0,
+        vertical: out.1,
+    })
+}
```

Comment:
Sorry to be a pain about this, but I'm sticking to my view that this doesn't belong in a `las` crate, as these are general operations not specific to the **las** specification. _To me_, it makes sense either to roll this operations into some general-purpose CRS crate, or make a small standalone conversion crate.